File: src/main/java/com/zepben/awsutils/S3.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.awsutils;

import com.amazonaws.SdkClientException;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.amazonaws.services.s3.model.PutObjectResult;
import com.zepben.annotations.EverythingIsNonnullByDefault;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.function.Supplier;

@EverythingIsNonnullByDefault
@SuppressWarnings({"WeakerAccess", "UnusedReturnValue"})
public class S3 {

    private final AmazonS3 s3Client;

    public S3(Dependencies dependencies) {
        this.s3Client = dependencies.s3ClientSupplier().get();
    }

    public PutObjectResult putObject(String bucketName, String key, String content) throws SdkClientException {
        ObjectMetadata metadata = new ObjectMetadata();

        int index = key.lastIndexOf('.');
        if ((index > 0) && (index < key.length() - 1))
            metadata.setContentType("text/" + key.substring(index + 1));
        else
            metadata.setContentType("text/plain");

        metadata.setContentLength(content.length());

        return putObject(bucketName, key, new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8)), metadata);
    }

    public PutObjectResult putObject(String bucketName, String key, InputStream content, ObjectMetadata metadata) throws SdkClientException {
        return s3Client.putObject(bucketName, key, content, metadata);
    }

    public interface Dependencies {

        Supplier<AmazonS3> s3ClientSupplier();

    }

}

File: src/main/kotlin/com/zepben/awsutils/S3.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.awsutils

import com.amazonaws.SdkClientException
import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.AmazonS3ClientBuilder
import com.amazonaws.services.s3.model.ObjectMetadata
import com.amazonaws.services.s3.model.PutObjectResult
import java.io.ByteArrayInputStream
import java.io.InputStream

class S3(
    private val s3Client: AmazonS3 = AmazonS3ClientBuilder.defaultClient()
) {

    @Throws(SdkClientException::class)
    fun putObject(bucketName: String, key: String, content: String): PutObjectResult {
        val metadata = ObjectMetadata().apply {
            contentType = key.toContentType()
            contentLength = content.length.toLong()
        }

        return putObject(bucketName, key, ByteArrayInputStream(content.toByteArray()), metadata)
    }

    @Throws(SdkClientException::class)
    fun putObject(bucketName: String, key: String, content: InputStream, metadata: ObjectMetadata): PutObjectResult =
        s3Client.putObject(bucketName, key, content, metadata)

    private fun String.toContentType(): String {
        // We add 1 to the index of the last `.` as we only want the extension.
        val index = lastIndexOf('.') + 1

        // We do not want to process items when the last `.` is actually the first character. i.e. "hidden" files.
        val type = if (index in 2..<length) substring(index) else "plain"

        return "text/$type"
    }

}
